### DISCUSSION TITLE
How to mock "redirect to file"?

### DISCUSSION THREAD

**Discussion by Andy-AO** [pester/Pester#2075] [Q&A] [answered]
Wanted to mock `>` and `>>`, mock `Out-File` is useful, but fails to get any arguments.

Is there a better solution other than replacing `>` with `Out-File`?

```PowerShell
Describe 'root' {
    it 'Symbol' {
        { mock '>' } | Should -Throw '*Could not find Command >*'
    }
    it 'Out-File' {
        function redirect {
            'text' > 'MyFake:\Path\Dir'
        }
        { redirect } | Should -Throw '*Cannot find drive*'
        mock 'Out-File' {
            $InputObject
            $LiteralPath
            $FilePath
            $args
        }
        # Can mock successfully, but cannot get any parameters.
        (redirect) | Should -Be $null
        Should -Invoke 'Out-File' -Exactly 1
    }
}

```

**Comment by fflaten** [comment, marked as answer]
Redirect operators aren't commands, but features in the language engine. You can't mock them. Mocking only works for commands that you can see with `Get-Command`.

Can you say more about why you would need this?

**Reply by Andy-AO** [reply]
mock `>` just as mock `Out-File` does.
It is possible to mock. If you look at the test code carefully, you will find that the mock has been successful, but the parameters cannot be used.

**Reply by fflaten** [reply]
Sorry, I didn't have time to test it test it. I knew it shared some code, but TIL that it [actually invokes `out-file`](https://github.com/PowerShell/PowerShell/blob/17986d88c10eb6271b1aad2ca955889e45e2b285/src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs#L1167-L1167). The mock is invoked with text and path set as `$InputObject` and `$FilePath` which you could see using Pester's Diagnostics output. 

However as I mentioned this operator is a engine feature, so the cmdlet isn't invoked the normal way. It's called in a special pipelineprocessor in which they throw away the output. That's most likely done because a redirection operator is never expected to output anything, but in case it does they stop it from polluting your remaining pipeline (+ better performance I suppose).

I'm not familiar with the PowerShell-codebase, but maybe one of the contributors in https://github.com/PowerShell/PowerShell can explain it if you need more details. It's just how the engine works, so nothing we can fix in Pester. 🙂

**Reply by Andy-AO** [reply]
@fflaten 
(╯▽╰) So it seems that we can only change our usage habits.
Thank you for your reply.
